### ISSUE TITLE
Exit Handler "hook" feature causes pod annotations to exceed 256KB

### ISSUE THREAD

**Issue by ad22** [argoproj/argo-workflows#6216]
## Summary 
When using exit handler hooks in a template, I see the following error on the wait container
`time="2021-06-24T06:15:03.562Z" level=warning msg="Output is larger than the maximum allowed size of 256 kB, only the last 256 kB were saved"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving output parameters"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving path output parameter: podname"
time="2021-06-24T06:15:03.562Z" level=info msg="Copying /podinfo/podname from volume mount"
time="2021-06-24T06:15:03.562Z" level=info msg="Successfully saved output parameter: podname"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving path output parameter: podnamespace"
time="2021-06-24T06:15:03.562Z" level=info msg="Copying /podinfo/podnamespace from volume mount"
time="2021-06-24T06:15:03.562Z" level=info msg="Successfully saved output parameter: podnamespace"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving path output parameter: poduid"
time="2021-06-24T06:15:03.562Z" level=info msg="Copying /podinfo/poduid from volume mount"
time="2021-06-24T06:15:03.562Z" level=info msg="Successfully saved output parameter: poduid"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving path output parameter: nodename"
time="2021-06-24T06:15:03.562Z" level=info msg="Copying /m/nodename from volume mount"
time="2021-06-24T06:15:03.562Z" level=info msg="Successfully saved output parameter: nodename"
time="2021-06-24T06:15:03.562Z" level=info msg="Saving path output parameter: image"
time="2021-06-24T06:15:03.562Z" level=info msg="No output artifacts"
time="2021-06-24T06:15:03.562Z" level=info msg="Annotating pod with output"
time="2021-06-24T06:15:03.653Z" level=info msg="Patch pods 422"
time="2021-06-24T06:15:03.654Z" level=warning msg="Non-transient error: Pod \"smoke-anish.dangi-df6256e-l2qp7sr-917260245\" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"
time="2021-06-24T06:15:03.654Z" level=error msg="executor error: Pod \"smoke-anish.dangi-df6256e-l2qp7sr-917260245\" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"
time="2021-06-24T06:15:03.654Z" level=info msg="Killing sidecars []"
time="2021-06-24T06:15:03.654Z" level=info msg="Alloc=16640 TotalAlloc=24257 Sys=76625 NumGC=9 Goroutines=9"
time="2021-06-24T06:15:03.654Z" level=fatal msg="Pod \"smoke-anish.dangi-df6256e-l2qp7sr-917260245\" is invalid: metadata.annotations: Too long: must have at most 262144 bytes"`

What happened/what you expected to happen?
Exit Handler runs successfully
## Diagnostics

👀 Yes! We need all of your diagnostics, please make sure you add it all, otherwise we'll go around in circles asking you for it:
Exit Handler Spec
``` templates:
  - container:
      args:
      - -c
      - set -xe && curl -O http://archive/jenkins_scripts/k8s_logger.py
        && python k8s_logger.py
      command:
      - bash
      image: '{{inputs.parameters.image}}'
      volumeMounts:
      - mountPath: /var/log/pods
        name: kubelet-pod-store
        readOnly: true
      - mountPath: /mnt/fio/docker-store/containers
        name: docker-container-store-linux
        readOnly: true
      - mountPath: /m
        name: workspace
    initContainers:
    - args:
      - -c
      - set -xe && pod_path=/var/log/pods/{{inputs.parameters.podnamespace}}_{{inputs.parameters.podname}}_{{inputs.parameters.poduid}}
        && apt install -y jq && find $pod_path | grep log | xargs -i cat {} | jq -r
        "[.time, .stream, .log] | @tsv" >> /m/concat.out && cat /m/concat.out
      command:
      - bash
      image: '{{inputs.parameters.image}}'
      mirrorVolumeMounts: true
      name: concat
    inputs:
      parameters:
      - name: podname
      - name: podnamespace
      - name: poduid
      - name: nodename
      - name: image
      - name: podresult
    name: LinuxExitHandler
    nodeSelector:
      kubernetes.io/hostname: '{{inputs.parameters.nodename}}'
```

Sample hooks spec referencing exit handler in a DAG node
```
spec:
......
        hooks:
          exit:
            arguments:
              parameters:
              - name: podname
                value: '{{tasks.Release-u3-base.outputs.parameters.podname}}'
              - name: nodename
                value: '{{tasks.Release-u3-base.outputs.parameters.nodename}}'
              - name: podnamespace
                value: '{{tasks.Release-u3-base.outputs.parameters.podnamespace}}'
              - name: poduid
                value: '{{tasks.Release-u3-base.outputs.parameters.poduid}}'
              - name: image
                value: '{{tasks.Release-u3-base.outputs.parameters.image}}'
              - name: podresult
                value: '{{tasks.Release-u3-base.outputs.result}}'
            template: LinuxExitHandler
```

What Kubernetes provider are you using? 
Bare Metal

What version of Argo Workflows are you running? 
3.1.0

What executor are you running? Docker/K8SAPI/Kubelet/PNS/Emissary
Emissary

Did this work in a previous version? I.e. is it a regression?
This has been working in a previous version of the argoexec image (don't know which one)
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
@sarabala1979 I can see two solutions to this. 

1. Compress the annotation using gzip and then base 122 (like base 64, but better compression ratio). Needs to be backwards compatible.
2. Use `taskset/status` for these.

**Comment by ad22** [comment]
This was caused by the the last parameter(podresult) `value: '{{tasks.Release-u3-base.outputs.result}}'`
"result" stores the entire stdout of the pod, and the default trim size is 256K which leaves no space for any other parameters. I've removed it and am using a workaround for my use case (using the docker container config.v2.json on the node directly to determine pod pass/fail status), and the issue no longer appears.

To provide some context - I'm trying to run the exit handler pod on the same node as the main pod that completed, via passing down nodename. Passing podresult was done only to obtain the status of the pod, not stdout. The exit handler will concat and store the all the pod container logs and result to an external store. It does so by accessing /var/lib/kubelet and docker container dir mounted into the handler pod. This is because I do not want to pass around the entire console log over the network in general, and especially as annotations, which is unscalable.
